File: packages/baran-core/src/main/kotlin/baran/geo/PlusCode.kt
package baran.geo

import kotlin.math.floor

/**
 * Minimal Open Location Code (Plus Code) codec — pair precision only.
 * 10-digit codes resolve to ~14 m, 8-digit (coarse / sensitive) to ~275 m.
 * Used for SOS location capture and the offline map. No network, no dependencies.
 */
object PlusCode {
    private const val ALPHABET = "23456789CFGHJMPQRVWX"

    data class LatLng(val lat: Double, val lng: Double, val cell: Double)

    /** Encode to a full plus code (e.g. "77GR2J4C+9P"). codeLen must be even, 2..10. */
    fun encode(lat: Double, lng: Double, codeLen: Int = 10): String {
        require(codeLen in 2..10 && codeLen % 2 == 0) { "codeLen must be even 2..10" }
        var aLat = (lat.coerceIn(-90.0, 89.9999999) + 90.0)
        var aLng = (normalizeLng(lng) + 180.0)
        var res = 20.0
        val sb = StringBuilder()
        repeat(codeLen / 2) {
            val latIdx = floor(aLat / res).toInt().coerceIn(0, 19)
            val lngIdx = floor(aLng / res).toInt().coerceIn(0, 19)
            sb.append(ALPHABET[latIdx]); sb.append(ALPHABET[lngIdx])
            aLat -= latIdx * res
            aLng -= lngIdx * res
            res /= 20.0
            if (sb.length == 8) sb.append('+')
        }
        if (!sb.contains('+')) sb.append('+') // short codes still carry the separator
        return sb.toString()
    }

    /** Coarsen to the 8-char (sensitive) cell, mirroring contract `plus_code8`. */
    fun coarse(code: String): String {
        val noPlus = code.replace("+", "")
        return if (noPlus.length >= 8) noPlus.substring(0, 8) else noPlus
    }

    /** Decode a full or coarse code to its cell CENTER. Returns null if invalid. */
    fun decode(raw: String?): LatLng? {
        if (raw == null) return null
        val code = raw.replace("+", "").trimEnd('0').uppercase()
        if (code.length < 2) return null
        var lat = -90.0
        var lng = -180.0
        var res = 20.0
        var lastRes = 20.0
        var i = 0
        while (i + 1 <= code.length && i < 10) {
            val a = ALPHABET.indexOf(code[i])
            val b = ALPHABET.indexOf(code[i + 1])
            if (a < 0 || b < 0) return null
            lat += a * res
            lng += b * res
            lastRes = res
            res /= 20.0
            i += 2
        }
        return LatLng(lat + lastRes / 2, lng + lastRes / 2, lastRes)
    }

    private fun normalizeLng(lng: Double): Double {
        var v = lng
        while (v < -180.0) v += 360.0
        while (v >= 180.0) v -= 360.0
        return v
    }
}

File: apps/coordinator-web/src/lib/plus-code.ts
// Minimal offline Open Location Code (Plus Code) decoder: full or coarse (8-char)
// code → cell CENTER {lat,lng}. ~14 m at 10 digits / ~275 m at 8 digits, no basemap tiles.
const ALPHABET = '23456789CFGHJMPQRVWX';

export interface LatLng {
  lat: number;
  lng: number;
  /** approximate cell size in degrees (for jitter/precision display) */
  cell: number;
}

export function decodePlusCode(raw?: string | null): LatLng | null {
  if (!raw) return null;
  const code = raw.replace('+', '').replace(/0+$/, '').toUpperCase();
  if (code.length < 2) return null;

  let lat = -90;
  let lng = -180;
  let res = 20; // degrees per unit at the current pair
  let lastRes = 20;
  let i = 0;
  for (; i + 1 < code.length + 1 && i < 10 && i + 1 <= code.length; i += 2) {
    const latDigit = ALPHABET.indexOf(code[i]);
    const lngDigit = ALPHABET.indexOf(code[i + 1]);
    if (latDigit < 0 || lngDigit < 0) return null;
    lat += latDigit * res;
    lng += lngDigit * res;
    lastRes = res;
    res /= 20;
  }
  // center of the last decoded cell
  return { lat: lat + lastRes / 2, lng: lng + lastRes / 2, cell: lastRes };
}

/** Encode lat/lng to a full Plus Code (pair precision). Inverse of decodePlusCode. */
export function encodePlusCode(lat: number, lng: number, codeLen = 10): string {
  let nlng = lng;
  while (nlng < -180) nlng += 360;
  while (nlng >= 180) nlng -= 360;
  let aLat = Math.min(Math.max(lat, -90), 89.9999999) + 90;
  let aLng = nlng + 180;
  let res = 20;
  let s = '';
  for (let i = 0; i < codeLen / 2; i++) {
    const latIdx = Math.min(19, Math.max(0, Math.floor(aLat / res)));
    const lngIdx = Math.min(19, Math.max(0, Math.floor(aLng / res)));
    s += ALPHABET[latIdx] + ALPHABET[lngIdx];
    aLat -= latIdx * res;
    aLng -= lngIdx * res;
    res /= 20;
    if (s.length === 8) s += '+';
  }
  if (!s.includes('+')) s += '+';
  return s;
}

/** Coarse (sensitive) 8-char cell from a fuller code, mirroring the contract. */
export function toPlusCodeCoarse(code?: string | null): string | null {
  if (!code) return null;
  const noPlus = code.replace('+', '');
  return noPlus.length >= 8 ? noPlus.substring(0, 8) : noPlus;
}
